src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/DF_BaseRecalibrator.java

```diff
@@ -0,0 +1,310 @@
+import org.broadinstitute.hellbender.engine.FeatureDataSource;
+import org.broadinstitute.hellbender.engine.FeatureInput;
+import org.broadinstitute.hellbender.engine.dataflow.ReadsSource;
+import org.broadinstitute.hellbender.engine.filters.ReadFilter;
+import org.broadinstitute.hellbender.tools.recalibration.RecalibrationTables;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+
+import java.io.*;
+import java.nio.channels.Channels;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * First pass of the base quality score recalibration -- Generates recalibration table based on various covariates
+ * (such as read group, reported quality score, machine cycle, and nucleotide context).
+ * <p>
+ * Dataflow version of BaseRecalibrator.
+ */
+@CommandLineProgramProperties(
+        usage = "First pass of the Base Quality Score Recalibration (BQSR) -- Generates recalibration table based on various user-specified covariates (such as read group, reported quality score, machine cycle, and nucleotide context).",
+        usageShort = "Generates recalibration table",
+        programGroup = ReadProgramGroup.class
+)
+public class DF_BaseRecalibrator extends CommandLineProgram {
+
+    public static final String APP_NAME = "Hellbender";
+
+    // ------------------------------------------
+    // Command-line options
+
+    /**
+     * all the command line arguments for BQSR and its covariates
+     */
+    @ArgumentCollection(doc = "all the command line arguments for BQSR and its covariates")
+    private transient final BaseRecalibrationArgumentCollection BRAC = new BaseRecalibrationArgumentCollection();
+
+    /**
+     * Path to the input (BAM file). Eventually can be local or on GCS.
+     */
+    protected String beforePath = null;
+
+
+    /**
+     * Path to the reference (SAM file). Eventually can be filesystem or gs://
+     */
+    protected String referencePath = null;
+
+    /**
+     * Path to save the output to. Must be local for now.
+     */
+    @Argument(doc = "Path to save the output to. Must be filesystem.", optional = false)
+    protected String outputTablesPath = "df-recaltables.sj";
+
+    // ------------------------------------------
+    // Dataflow-specific options
+
+    @Argument(doc = "path to client-secrets.json", optional = false)
+    protected String secretsFile;
+
+    // not actually optional if you ask for remote exec
+    @Argument(doc = "Google Cloud project name", optional = true)
+    protected String project;
+
+    // not actually optional if you ask for remote exec
+    @Argument(doc = "Google Cloud Storage path for intermediate files", optional = true)
+    protected String stagingLocation;
+
+    @Argument(doc = "Number of Dataflow workers", optional = true)
+    protected int numWorkers = 0;
+
+    // ------------------------------------------
+
+    // A holder for the inputs to BQSR
+    public static class BQSRInputs {
```

Comment:
It's certainly possible to get rid of this struct and use multiple member variables instead. This grouping was made to put related objects together, but since my goal is to be maximally useful to you, I can adjust to the style you prefer.